### PR TITLE
Modeling issue 31 naive bayes

### LINKED ISSUE #31: Modeling with Naive Bayes
* Train a Naive Bayes model for all labels.
* Use a grid search and MLFlow.
* Possible Preprocessing: Tfidf, WordCount

### PR DESCRIPTION
**Pull request by lima-tango**
Done:
- Add logging of confusion matrix
- Fix utils.feature_engineering.add_column_text
- Add cycle tag
- Fix error with NaN due to new train-test-val-split
  + Move iteration for labels from run_training() into __main__
  + Move __get_data() call from run_training() into __main__

Needs improvement:
- Slow due to repeated loading call that takes 14s
- Long logging output due to stopwords

Closes #31

### CODE REVIEW

**Review by dominikmn — COMMENTED** [review]
Good work :+1: 
### Specific points
I have listed down my *specific* comments in the code.

### General point
Further below, I mentioned the **Option "Sharing as object oriented attribute"**.
In general, an *object-oriented* version of the modeling_template will make it easier to share data between the different functions like `run_training`, `__get_data` and an optional *evaluation function*. I mean data like: 
* the dataframes (`X_train`, `y_val`, etc.)
* `grid_search_params`
* `mlflow_params`
* mlflow sessions

However, this would definitely exceed the scope of a "quick win" and I wonder how much we would really benefit from it during the remaining time of the project.

File: modeling/naive_bayes.py [hunk 1]
```diff
@@ -51,35 +52,37 @@
...
-           'label_offtopic', 'label_personalstories', 'label_possiblyfeedback',
-           'label_sentimentnegative', 'label_sentimentneutral', 'label_sentimentpositive',]
     X_train = df_train.text
-    y_train = df_train[target_labels]
+    y_train = df_train[target_label]
     X_val = df_val.text
-    y_val = df_val[target_labels]
-    return X_train, X_val, y_train, y_val
+    y_val = df_val[target_label]
+    return X_train, y_train, X_val, y_val
```
Comment (dominikmn) [inline modeling/naive_bayes.py:79]: We could speed up the training by 'caching' the dataframes on the disk:
```
try:
        df_posts_train = pd.read_csv(f'./cache/{target_label}_train.csv', sep='\t')
        df_posts_val =pd.read_csv(f'./cache/{target_label}_val.csv', sep='\t')
except:
        df_posts_train = loading.load_extended_posts(split="train", label=target_label)
        df_posts_val = loading.load_extended_posts(split="val", label=target_label)

        df_train = feature_engineering.add_column_text(df_posts_train)
        df_val = feature_engineering.add_column_text(df_posts_val)

        df_train = df_train[['text', target_label]]
        df_val = df_val[['text', target_label]]

        df_train.to_csv(f'./cache/{target_label}_train.csv', sep='\t')
        df_val.to_csv(f'./cache/{target_label}_val.csv', sep='\t')
```

**Review by dominikmn — APPROVED** [review]

File: modeling/naive_bayes.py [hunk 2]
```diff
@@ -134,7 +137,7 @@ def predict_with_threshold(y_pred_proba: np.array, threshold: float) -> np.array
 
 
 # %%
-def run_training(model_details, mlflow_params) -> None:
+def run_training(X_train, y_train, X_val, y_val, model_details, mlflow_params, mlflow_tags) -> None:
```
[the remark this answers is not in the document]
Comment (lima-tango) [reply modeling/naive_bayes.py:140]: Thank you for the input! Handled in Issue #59

**Comment by lima-tango** [comment]
This MR will not be merged. The suggested improvements are discussed/included in #62. A new MR will be written to solve Issue #31.

**Pull request closed without merging**